File: GaussMPI.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>

int size; //размер матрицы
FILE* fp1;
FILE* fp2;

int main(int argc, char* argv[])
{
	int size = 10; 
	int rank, numprocs;
	double* m = (double*)malloc(sizeof(*m) * (size) * (size + 1));
	int i, j;

	fp1 = fopen("m.txt", "r");
	for (i = 0; i < size; i++) {
		for (j = 0; j < size + 1; j++)
		{
			fscanf(fp1, "%lf", &m[i * (size + 1) + j]);
			//printf("%lf", m[i * (size + 1) + j]);
		}
		//printf("\n");
	}
	fclose(fp1);

	MPI_Init(NULL, NULL);
	double time = MPI_Wtime();
	MPI_Comm_rank(MPI_COMM_WORLD, &rank); //уникальный номер - ранг процесса
	MPI_Comm_size(MPI_COMM_WORLD, &numprocs); //общее количество доступных процессов

	//количество строк для процесса
	int tmp = size / numprocs;
	if (size % numprocs) 
		tmp++; 
	int rest = numprocs * tmp - size;
	int numrows = tmp;
	if (rank >= numprocs - rest) //случай последнего процесса, которому достается остаток строк
		numrows = tmp - 1;

	int* rows_number = (int*)malloc(sizeof(*rows_number) * numrows); //номера строк, которые должен передать оставшимся процессам конкретный процесс

	double* a = (double*)malloc(sizeof(*a) * numrows * (size + 1)); //коэффициенты со столбцом свободных членов
	double* x = (double*)malloc(sizeof(*x) * size); //решения системы
	double* t = (double*)malloc(sizeof(*t) * (size + 1));

	for (i = 0; i < numrows; i++) {
		rows_number[i] = rank + numprocs * i; //заносим номера строк матрицы
		//printf("num: %d; ", rows_number[i]);
		for (j = 0; j < size + 1; j++) {
			a[i * (size + 1) + j] = m[rows_number[i] * (size + 1) + j];
			//printf("%lf", a[rows_number[i] * (size + 1) + j]);
		}
		//a[i * (size + 1) + size] = m[rows_number[i] * (size + 1) + (size)];
		//printf("%lf", a[rows_number[i] * (size + 1) + size]);
		//printf("\n");
	}

	//Elimination
	int r = 0;
	for (i = 0; i < size - 1; i++) {
		// Убираем из последующих уравнений xi
		if (i == rows_number[r]) {
			// Посылаем в последующие уравнения строку i 
			MPI_Bcast(&a[r * (size + 1)], size + 1, MPI_DOUBLE, rank, MPI_COMM_WORLD);
			for (j = 0; j <= size; j++)
				t[j] = a[r * (size + 1) + j];
			r++;
		}
		else {
			MPI_Bcast(t, size + 1, MPI_DOUBLE, i % numprocs, MPI_COMM_WORLD);
		}
		// Принятая строка вычитается из тех уравнений, за которые отвечает данный процесс
		for (j = r; j < numrows; j++) {
			double scaling = a[j * (size + 1) + i] / t[i];
			for (int k = i; k < size + 1; k++)
				a[j * (size + 1) + k] -= scaling * t[k];
		}
	}

	//Заполняем вектор x
	r = 0;
	for (i = 0; i < size; i++) {
		x[i] = 0;
		if (i == rows_number[r]) {
			x[i] = a[r * (size + 1) + size];
			r++;
		}
	}
	//Back substitution
	r = numrows - 1;
	for (i = size - 1; i > 0; i--) {
		if (r >= 0) {
			if (i == rows_number[r]) {
				x[i] /= a[r * (size + 1) + i];
				// Передаем найденное x_i
				MPI_Bcast(&x[i], 1, MPI_DOUBLE, rank, MPI_COMM_WORLD);
				r--;
			}
			else
				MPI_Bcast(&x[i], 1, MPI_DOUBLE, i % numprocs, MPI_COMM_WORLD);
		}
		else
			MPI_Bcast(&x[i], 1, MPI_DOUBLE, i % numprocs, MPI_COMM_WORLD);
		for (int j = 0; j <= r; j++)
			x[rows_number[j]] -= a[j * (size + 1) + i] * x[i];
	}
	if (rank == 0)
		x[0] /= a[r * (size + 1)];
	MPI_Bcast(x, 1, MPI_DOUBLE, 0, MPI_COMM_WORLD); //все процессы содержат вектор решений
	free(t);
	free(rows_number);
	free(a);
	time = MPI_Wtime() - time;
	
	//запись результатов в файл
	fp2 = fopen("x_mpi_c.txt", "w");
	for (i = 0; i < size; i++) {
		fprintf(fp2, "%lf ", x[i]);
	}
	fclose(fp2);

	if (rank == 0) {
		printf("Time: %.6fs\n", time);
		printf("MPI/C: size of matrice %d, processes %d\n", size, numprocs);
	}

	free(x);
	MPI_Finalize();
	return 0;
}

File: GaussMPI.py
from mpi4py import MPI
import time
import numpy as np

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
numprocs = comm.Get_size()

size = 10
mtrxA = np.array([])

t = np.array([])
t = [0] * (size+2)
vecX = np.array([])
vecX = [0] * (size)

def elimination(data, r):
    #print("rank", r)
    start = int(data[-1][size+1])
    for k in range(len(data)-1):
        scaling = data[k][start] / data[-1][start]
        for l in range(start, size):
            data[k][l] -= data[-1][l] * scaling
        data[k][size] -= scaling * data[-1][size]
    data = np.delete(data, -1, 0)
    #print("after elim", data)
    return data 
       
def init_matrice():
    global mtrxA
    global B
    mtrxA = np.loadtxt("m.txt")
    #print(mtrxA)

if __name__ == '__main__':
    if (rank == 0):
        init_matrice()

        start_time = time.time()
                       
        for row_n in range(size - 1):
            newA = np.array([])
            data = np.array([])
            data_add = mtrxA[0:row_n+1][:]
            rows_number_add = np.array([]) #строки матрицы до row_n включительно
            
            for i in range(row_n+1):
                rows_number_add = np.append(rows_number_add, row_n)
            data_add = np.hstack((data_add, np.atleast_2d(rows_number_add).T))
            
            numpr = numprocs #номер процесса, начиная с которого data будет пустым
            
            for num in range(numprocs):
                data = mtrxA[row_n+num+1:size:numprocs][:]
                data = np.vstack([data, mtrxA[row_n][:]]) #добавили строку row_n
                if (len(data) > 1): #если есть данные, которые надо изменять
                    rows_number = np.array([]) #номера строк в data (доп.столбец)
                    for i in range(len(data) - 1):
                        rows_number = np.append(rows_number, (row_n+num+1)+i*numprocs)
                    rows_number = np.append(rows_number, row_n)
                    #print("newdata", data)
                    data = np.hstack((data, np.atleast_2d(rows_number).T))
                else: 
                    if (numpr == numprocs):
                        numpr = num
                    
                #print("data", data)
                
                if (num > 0):
                    comm.send(data, dest=num, tag = row_n)
                else: #работа 0-го процесса
                    if (len(data) > 1):
                        new_data = elimination(data, rank)
                        newA = new_data  
                      
            for n in range(1, numpr):
                new_data = comm.recv(source=n, tag=n)
                newA = np.append(newA, new_data, 0)
            newA = np.append(newA, data_add, 0)
            newA = newA[newA[:, -1]. argsort()]  #сортировка по изначальным номерам строк
            newA = np.delete(newA, -1, 1)
            mtrxA = newA
        #print("mtrxA", mtrxA)    
        
        #Back substitution
        for row_i in range(size - 1, -1, -1): 
            vecX[row_i] = mtrxA[row_i][size]
            for row_j in range(size - 1, row_i, -1):
                vecX[row_i] -= mtrxA[row_i][row_j] * vecX[row_j]
            vecX[row_i] = vecX[row_i] / mtrxA[row_i][row_i]
        #print(vecX)
        
    if (rank > 0):
        for i in range(size - 1):
            data = comm.recv(source=0, tag = i)
            if (len(data) > 1):
                new_data = elimination(data, rank)
                comm.send(new_data, dest=0, tag=rank) 
               
    if (rank == 0):
        np.savetxt('x_mpi_py.txt', vecX, fmt='%10.5f')
        end_time = time.time()
        print('Time:{}s'.format(end_time - start_time))
        print('MPI/Python: size of matrice', size, ', processes', numprocs)
        


